File: highest value palindrome /Result.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;
import java.util.regex.*;
import java.util.stream.*;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;

class Result {
  
    public static String highestValuePalindrome(String s, int n, int k) {
    // Write your code here
        char[] c= s.toCharArray();
        
        //Try to make s a palindrome firtst
        for(int i=0; i<= n/2; i++)
        {
            int j= n-i-1;
            if(c[i] != c[j]){
                if(k<= 0)
                {
                    return "-1";
                }
                char max= c[i] > c[j] ? c[i] : c[j];
                c[i]= max;
                c[j]= max;
                k--;
            }
        }

        for(int i=0; k>0 && i<= n/2; i++)
        {                    System.out.println(k);

            int j= c.length-i-1;
            if( c[i]=='9' || c[j]=='9')
                continue;
            //If you changed a character to value !=9, undo, change it to 9, 
            //and change the corresponding character to 9 with one additional move  
            if(c[i] != s.charAt(i) || c[j]!= s.charAt(j) || i == j)
            {
                c[i]= '9';
                c[j]= '9';
                k--;
            }else{
                if(k>=2)
                {
                    c[i]='9';
                    c[j]= '9';
                    k-= 2;
                }
            }
            
        }
        return String.valueOf(c);
    }

}

public class Solution {
    public static void main(String[] args) throws IOException {
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        String[] firstMultipleInput = bufferedReader.readLine().replaceAll("\\s+$", "").split(" ");

        int n = Integer.parseInt(firstMultipleInput[0]);

        int k = Integer.parseInt(firstMultipleInput[1]);

        String s = bufferedReader.readLine();

        String result = Result.highestValuePalindrome(s, n, k);

        bufferedWriter.write(result);
        bufferedWriter.newLine();

        bufferedReader.close();
        bufferedWriter.close();
    }
}

File: highest value palindrome /Result.py
#!/bin/python3

import math
import os
import random
import re
import sys

#
# Complete the 'highestValuePalindrome' function below.
#
# The function is expected to return a STRING.
# The function accepts following parameters:
#  1. STRING s
#  2. INTEGER n
#  3. INTEGER k
#


def highestValuePalindrome(s, n, k):
    # Write your code here

    # two pointers left and right for traversing from both ends
    left = 0
    right = len(s) - 1

    result = list(s[::])

    while left <= right and k >= 0:
        # if not equal than replace the value with the maximum one
        if result[left] != result[right]:
            result[left] = max(result[left], result[right])
            result[right] = max(result[left], result[right])
            k -= 1

        left += 1
        right -= 1

    # it means that we can make above string palindrome within the max limit of changes allowed
    if k < 0:
        return '-1'

    left = 0
    right = len(s) - 1

    while left <= right and k > 0:
        # string is of odd length and we are at center
        if left == right and k > 0:
            result[left] = result[right] = "9"
            k -= 1
            continue

        # if current character is less than 9 then we can replace it with 9 if swaps are available
        if result[left] < "9":
            if k > 0 and result[left] != s[left] or result[right] != s[right]:
                # if we have already swap that index then we can ignore that and replace its value with 9
                result[left] = result[right] = "9"
                # only single swap is needed because we discarded the previous swap
                k -= 1
            elif k >= 2 and result[left] != result[right]:
                # if they are not previously swapped then replaced both with 9 and dec by 2 because we have done 2 swaps
                result[left] = result[right] = "9"
                k -= 2

        left += 1
        right -= 1

    return "".join(result)


if __name__ == "__main__":
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    first_multiple_input = input().rstrip().split()

    n = int(first_multiple_input[0])

    k = int(first_multiple_input[1])

    s = input()

    # n = 4
    # k = 1
    # s = "3043"

    result = highestValuePalindrome(s, n, k)

    # print(result)

    fptr.write(result + '\n')

    fptr.close()
